src/magic_ter_fur_transform.cpp

```diff
@@ -0,0 +1,287 @@
+#include "magic.h"
+
+#include "creature.h"
+#include "point.h"
+#include "game.h"
+#include "generic_factory.h"
+#include "map.h"
+#include "mapdata.h"
+#include "messages.h"
+#include "type_id.h"
+
+namespace
+{
+generic_factory<ter_furn_transform> ter_furn_transform_factory( "ter_furn_transform" );
+} // namespace
+
+template<>
+const ter_furn_transform &string_id<ter_furn_transform>::obj() const
+{
+    return ter_furn_transform_factory.obj( *this );
+}
+
+template<>
+bool string_id<ter_furn_transform>::is_valid() const
+{
+    return ter_furn_transform_factory.is_valid( *this );
+}
+
+void ter_furn_transform::load_transform( JsonObject &jo, const std::string &src )
+{
+    ter_furn_transform_factory.load( jo, src );
+}
+
+void ter_furn_transform::reset_all()
+{
+    ter_furn_transform_factory.reset();
+}
+
+bool ter_furn_transform::is_valid() const
+{
+    return ter_furn_transform_factory.is_valid( this->id );
+}
+
+void ter_furn_transform::check_consistency()
+{
+    return;
+}
+
+const std::vector<ter_furn_transform> &ter_furn_transform::get_all()
+{
+    return ter_furn_transform_factory.get_all();
+}
+
+static void load_ter_results( JsonObject &jsi, const std::string &json_key,
+                              weighted_int_list<ter_id> &list )
+{
+    if( jsi.has_string( json_key ) ) {
```

Comment:
You can't really use `ter_id` here, you must use `ter_str_id`.

Try out the following: put in *one* json file the definition of a `ter_furn_transform` that refers to some *new* terrain type and put the definition of that type *afterwards*. When loading the `ter_furn_transform`, the game is going to convert you terrain id string (the new terrain type) into an integer (the `ter_id`), but it does not know that terrain yet and will fail.

Using the `ter_str_id` bypasses that, it will load just the string and wont convert it.

The same for furniture.